test/ds/test_dynamic_shape_models_dynamo.py

```diff
@@ -0,0 +1,125 @@
+import torch_xla.debug.metrics as met
+
+# It enables us to run python implementations of CompositeAutogradImplicit ops.
+# CompositeAutogradImplicit means we don't have an explicit backward formula for an op instead an op is composed of a bunch of ops that do have backward formulas and combines this formulas is equivalent to differentiating the op explicitly.
+pd = torch._C._EnablePythonDispatcher()
+xla_dev = xm.xla_device()
+torch._dynamo.config.capture_dynamic_output_shape_ops = True
+
+
+class Feedforward(torch.nn.Module):
+
+  def __init__(self, input_size, hidden_size):
+    super().__init__()
+    self.input_size = input_size
+    self.hidden_size = hidden_size
+    self.fc1 = torch.nn.Linear(self.input_size, self.hidden_size)
+    self.fc1.weight.data.fill_(0.01)
+    self.fc1.bias.data.fill_(0.01)
+    self.relu = torch.nn.ReLU()
+    self.fc2 = torch.nn.Linear(self.hidden_size, 1)
+    self.fc2.weight.data.fill_(0.01)
+    self.fc2.bias.data.fill_(0.01)
+    self.sigmoid = torch.nn.Sigmoid()
+
+  def forward(self, x):
+    hidden = self.fc1(x)
+    relu = self.relu(hidden)
+    output = self.fc2(relu)
+    output = self.sigmoid(output)
+    return output
+
+
+class TestDynamicShapeModels(unittest.TestCase):
+  def test_benchmark_backwards_pass_with_dynamic_input(self):
+    num_features = 2
+    model = Feedforward(num_features, hidden_size=10)
+    model.compile()
+
+    for i in range(0, 100):
+        self.backwards_pass_with_dynamic_input(model, num_features)
+
+  def test_backward_pass_with_dynamic_input(self):
+    num_features = 2
+    model = Feedforward(num_features, hidden_size=10)
+    self.backwards_pass_with_dynamic_input(model, num_features)
+
+  def test_compiled_backward_pass_with_dynamic_input(self):
+    num_features = 2
+    model = Feedforward(num_features, hidden_size=10)
+    model.compile()
+    self.backwards_pass_with_dynamic_input(model, num_features)
+
+  def backwards_pass_with_dynamic_input(self, model, num_features):
+    num_test_samples = 5
+    criterion = torch.nn.BCELoss()
+    optimizer = torch.optim.SGD(model.parameters(), lr=1e-3)
+    optimizer.zero_grad()
+
+    # training
+    model.train()
+    x_training, y_training = self.create_dynamic_test_data(
+        num_test_samples, num_features)
+
+    y_pred = model(x_training)
+    loss = criterion(y_pred.squeeze(), y_training)
+    # Backpropagation.
+    loss.backward()
+    print('Finished training.')
+
+    # testing
+    model.eval()
+    with torch.no_grad():
+      x_test, y_test = self.create_dynamic_test_data(num_test_samples,
+                                                     num_features)
+      y_pred = model(x_test)
+      criterion(y_pred.squeeze(), y_test).item()
+    print('Test passed.')
+
+  def create_dynamic_test_data(self,
+                               num_test_samples,
+                               num_features,
+                               num_non_zeros=1):
+    x_test = torch.zeros(num_test_samples, num_features)
```

Comment:
need to move the input and the model to xla device?

Reply:
Maybe I should've given more context, or let me know if this isn't generally OK in this repo. Since the original test didn't work on CPU, I couldn't test this on XLA. I mostly wanted to commit this to try to reproduce the benchmark results you had here - https://github.com/pytorch/xla/blob/master/docs/dynamic_shape.md and see what inductor did, so that when XLA works, this test could be reproduced.

The current test_dynamic_shape_models.py file is skipped on CI/CD since it isn't on a TPU. Does that make sense or are these kinds of tests not allowed in this repo?

Reply:
I see. Fwiw, test_dynamic_shape_models.py is running in the post-submit TPU CI https://github.com/pytorch/xla/blob/49a6eec96a24f7f4030f20e2fae9a43bf72599f9/.github/workflows/tpu_ci.yml#L44. Though, there's a plan to make TPU CI presubmit soon.

> when XLA works, this test could be reproduced.

Could you elaborate this? What's not working in XLA now?

Reply:
Ahh I see, I was only testing on CPU and the original test skips on CPU:

https://github.com/pytorch/xla/blob/master/test/ds/test_dynamic_shape_models.py#L50

See https://github.com/pytorch/xla/issues/4298

Is there a way to test TPU CI? Otherwise as a follow up, can I try to get this test working on TPU with XLA and torch.compile() ?

Reply:
I tired testing this on TPU and get the following errors:

```
FAILED test/ds/test_dynamic_shape_models_dynamo.py::TestDynamicShapeModels::test_backward_pass_with_dynamic_input - RuntimeError: torch_xla/csrc/aten_xla_bridge.cpp:95 : Check failed: xtensor
FAILED test/ds/test_dynamic_shape_models_dynamo.py::TestDynamicShapeModels::test_benchmark_backwards_pass_with_dynamic_input - torch._dynamo.exc.InternalTorchDynamoError: 'torch._C._SymNode' object has no attribute 'shape_env'
FAILED test/ds/test_dynamic_shape_models_dynamo.py::TestDynamicShapeModels::test_compiled_backward_pass_with_dynamic_input - torch._dynamo.exc.InternalTorchDynamoError: 'torch._C._SymNode' object has no attribute 'shape_env'
FAILED test/ds/test_dynamic_shape_models_dynamo.py::TestDynamicShapeModels::test_forward_pass_dynamic_input_correctness - torch._dynamo.exc.InternalTorchDynamoError: 'torch._C._SymNode' object has no attribute 'shape_env'
```

So would it be better to:

1) Check this in as with XLA support and disable the test
2) Check this in as a passing test without XLA support
3) Don't check this in at all?

Thanks!

Reply:
afaict, we don't usually merge non torch_xla related changes into the repo. So I'd like to understand the intention here. If the intention is to merge the test and let it run in the CI to prevent regression, we should merge it. If the test serves as a toy example to experiment what/how inductor does, then I'm hesitant to merge it. Wdyt?

Reply:
The intention was, use it as a toy example such that `torch.compile(..)` == the existing test that works with XLA, and make it easier for others to reproduce rather than pulling another branch / having it get stale. If that doesn't fit in this repo, then it's ok to not merge.

That's also why I originally just had it without XLA support in the first PR so that it was running in CI to prevent regression. Does that make sense?

Reply:
> That's also why I originally just had it without XLA support in the first PR so that it was running in CI to prevent regression.

Sorry I forgot which first PR you are referring to. By "prevent regression", do you mean the regression of native pytorch dynamo on dynamic input?

Reply:
Ahh all good, sorry for my miscommunication. I meant the original version of this PR was prevent a regression of the native PyTorch dynamo on dynamic input.

Reply:
If it's purely testing pytorch feature on dynamic input, shouldn't it be added to the pytorch repo?

Reply:
Sorry, this is the untouched version so that it works with dynamo and the test can be updated as we get support for XLA to slowly transform this test to use XLA. Does that make sense?

But generally it sounds like that type of development isn't supported in the PyTorch/XLA repo? Kind of like experimental?

Reply:
I see. If it helps make the development work easier, feel free to merge. But please update the pr description about the intention of adding this script and this script come from pytorch (with a link?).